### ISSUE TITLE
Custom generic serializer crashing compiler on JVM IR backend

### ISSUE THREAD

**Issue by gohlinka2** [Kotlin/kotlinx.serialization#1680]
**Describe the bug**

When a custom generic `KSerializer` class is created and applied on a `val` in a `Serializable` class, the compiler crashes with `java.lang.AssertionError: No such value argument slot in IrConstructorCallImpl.`. See steps to reproduce for the specific code that crashes. Expand following spoiler for the full crash log.   

<details>
  <summary>Error output</summary>
  
```
e: java.lang.AssertionError: No such value argument slot in IrConstructorCallImpl: 1 (total=1)
	at org.jetbrains.kotlin.ir.expressions.IrMemberAccessExpressionKt.throwNoSuchArgumentSlotException(IrMemberAccessExpression.kt:66)
	at org.jetbrains.kotlin.ir.expressions.IrFunctionAccessExpression.putValueArgument(IrFunctionAccessExpression.kt:31)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.irInvoke(GeneratorHelpers.kt:211)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.irInvoke(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.irInvoke(GeneratorHelpers.kt:222)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.irInvoke(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.serializerInstance(GeneratorHelpers.kt:1099)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.serializerInstance(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.formEncodeDecodePropertyCall(GeneratorHelpers.kt:1307)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.formEncodeDecodePropertyCall(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.formEncodeDecodePropertyCall$default(GeneratorHelpers.kt:1297)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.serializeAllProperties(GeneratorHelpers.kt:1232)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.serializeAllProperties(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator$generateWriteSelfMethod$1.invoke(SerializableIrGenerator.kt:335)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator$generateWriteSelfMethod$1.invoke(SerializableIrGenerator.kt:281)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.contributeFunction(GeneratorHelpers.kt:70)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.contributeFunction(SerializableIrGenerator.kt:35)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.generateWriteSelfMethod(SerializableIrGenerator.kt:281)
	at org.jetbrains.kotlinx.serialization.compiler.backend.common.SerializableCodegen.generateSyntheticMethods(SerializableCodegen.kt:51)
	at org.jetbrains.kotlinx.serialization.compiler.backend.common.SerializableCodegen.generate(SerializableCodegen.kt:25)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator$Companion.generate(SerializableIrGenerator.kt:354)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializerClassLowering.lower(SerializationLoweringExtension.kt:58)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:36)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitClass(IrElementVisitorVoid.kt:44)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitClass(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlin.ir.declarations.IrClass.accept(IrClass.kt:55)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.acceptChildren(IrFileImpl.kt:84)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitElement(SerializationLoweringExtension.kt:32)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitPackageFragment(IrElementVisitorVoid.kt:30)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitPackageFragment(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:37)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:38)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt$runOnFileInOrder$1.visitFile(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.accept(IrFileImpl.kt:81)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:271)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtensionKt.runOnFileInOrder(SerializationLoweringExtension.kt:30)
	at org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationLoweringExtension.generate(SerializationLoweringExtension.kt:77)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.convertToIr$lambda-1(JvmIrCodegenFactory.kt:126)
	at org.jetbrains.kotlin.psi2ir.Psi2IrTranslator.generateModuleFragment(Psi2IrTranslator.kt:89)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.convertToIr(JvmIrCodegenFactory.kt:146)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.convertToIr$default(JvmIrCodegenFactory.kt:64)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.generateModule(JvmIrCodegenFactory.kt:59)
	at org.jetbrains.kotlin.codegen.KotlinCodegenFacade.compileCorrectFiles(KotlinCodegenFacade.java:35)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.generate(KotlinToJVMBytecodeCompiler.kt:321)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:113)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:56)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:169)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:92)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:412)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:112)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally(IncrementalCompilerRunner.kt:358)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally$default(IncrementalCompilerRunner.kt:300)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:160)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:81)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:607)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:96)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1658)
	at sun.reflect.GeneratedMethodAccessor176.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:357)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
  
</details>

**To Reproduce**

```kotlin
@Serializable
data class Parent(
    @Serializable(with = MapRowListAsMapSerializer::class)
    val childMap: Map<String, Child>
)

@Serializable
data class Child(val name: String)

/**
 * Converts between a [Map] and a [List] of [MapRow]s.
 */
private class MapRowListAsMapSerializer<T>(private val tSerializer: KSerializer<T>) : KSerializer<Map<String, T>> {

    override val descriptor: SerialDescriptor = ListSerializer(MapRow.serializer(tSerializer)).descriptor

    override fun deserialize(decoder: Decoder): Map<String, T> {
        val rows: List<MapRow<T>> = decoder.decodeSerializableValue(ListSerializer(MapRow.serializer(tSerializer)))
        return rows.map { row -> row.key to row.value }.toMap()
    }

    override fun serialize(encoder: Encoder, value: Map<String, T>) {
        val rows = value.map { MapRow(it.key, it.value) }
        return encoder.encodeSerializableValue(ListSerializer(MapRow.serializer(tSerializer)), rows)
    }

    @Serializable
    private data class MapRow<T>(val key: String, val value: T)
}
```

And try to run deserialization of `Parent` in a JVM environment, JUnit test in my case.

**Expected behavior**

The compiler doesn't crash and the generic serializer receives the type serializer. 

**Environment**
 - Kotlin version: 1.5.30
 - Library version: 1.2.2 // tried also 1.3.0-RC but doesn't solve the issue
 - Kotlin platforms: JVM, Native - complier crashes when running it in a JUnit test
 - Gradle version: 7.2
 - IDE version: Android Studio Arctic Fox | 2020.3.1 Patch 1

**Comment by gohlinka2** [comment]
Also, if anyone is having the same issue, here is the workaround I used in the end, if applicable for you:

Make `MapRowListAsMapSerializer` `abstract` and create on `object` implementation for every concrete case of its usage, so:

```
private object ChildMapSerializer: MapRowListAsMapSerializer<Child>(Child.serializer())
// any other objects for all cases you need
```

And apply it on the usage site:

```
@Serializable(with = ChildMapSerializer::class)
val childMap: Map<String, List<Child>>
```

Of course, this is not applicable if you have a large or arbitrary amount of types needed to cover.

**Comment by sandwwraith** [comment]
You can also try to refactor class to a `class MapRowListAsMapSerializer<T>(val keySer: KSerializer<String>, private val tSerializer: KSerializer<T>): KSerializer<Map<String, T>>` and simply don't use first argument.

Since it is serializer for `Map`, it is expected to have two arguments

Related: #1587

**Comment by shanshin** [comment]
Fixed along with #1587. 
The number of parameters in the constructor must be 0 or match the number of type parameters in a serializable type
